lib/server/services/qualityControlFlag/QcFlagService.js

```diff
@@ -53,6 +56,192 @@ class QcFlagService {
         return qcFlag;
     }
 
+    /**
+     * Create new instance of quality control flags for data pass
+     * @param {Partial<QcFlag>} parameters flag instance parameters
+     * @param {object} [relations] QC Flag Type entity relations
+     * @param {Partial<UserIdentifier>} [relations.user] user identifiers
+     * @param {number} [parameters.flagTypeId] flag type id
+     * @param {number} [parameters.runNumber] associated run's number
+     * @param {number} [parameters.dataPassId] associated dataPass' id
+     * @param {number} [parameters.dplDetectorId] associated dplDetector's id
+     * @return {Promise<QcFlag>} promise
+     * @throws {BadParameterError, NotFoundError}
+     */
+    async createForDataPass(parameters, relations = {}) {
+        const {
+            from: fromTime = null,
+            to: toTime = null,
+            comment,
+        } = parameters;
+        const {
+            user: { userId, externalUserId } = {},
+            flagTypeId,
+            runNumber,
+            dataPassId,
+            dplDetectorId,
+        } = relations;
+
+        if (fromTime && toTime && !(fromTime < toTime)) {
+            throw new BadParameterError('Parameter "to" timestamp must be greater than "from" timestamp');
+        }
+        return dataSource.transaction(async () => {
+            // Check user
+            const user = await getUserOrFail({ userId, externalUserId });
+
+            // Check associations
+            const dplDetector = await DplDetectorRepository.findOne({ where: { id: dplDetectorId } });
+            if (!dplDetector) {
+                throw new BadParameterError(`No DPL Detector with given id (${dplDetectorId})`);
+            }
+            if (!dplDetector.name || NON_QC_DETECTORS.has(dplDetector.name)) {
+                throw new BadParameterError(`Bad DPL Detector selection, QC flag creation for "${dplDetector.name}" is impossible`);
+            }
```

Comment:
I would simplify this to a common `!dplDetector?.name || NON_QC_DETECTORS.has(dplDetector.name)` and a global error message like "Invalid DPL detector"